Split UDisk2 permissions, Use disable-mnt to control data storage perms.

LaakkonenJussi opened the pull request:
[permissions] Use disable-mnt to control data storage perms, move udisk2 include. OMP#JOLLA-49

Block access to any media mount with disable-mnt in Base.permissions.
Grant permissions to all media when DataStorages permission is used,
including UDisks2 D-Bus.

Move the UDisks permission include away from Base.permissions. This is
to be included separately or via DataStorages permission.

rainemak commented:
Nice! Need to check in more detail but that goes tomorrow.

Please combine references to one single line:
```
[permissions] Use disable-mnt to control data storage perms, move udisks2 include. JB#53818 OMP#JOLLA-49
```

Tomin1 commented on permissions/DataStorages.permission:
```diff
@@ -5,3 +5,9 @@
 # x-sailjail-description = Removable media
 # x-sailjail-translation-key-long-description =
```
Please fill in these values (catalog, key-description, key-long-description).

LaakkonenJussi replied:
Ah, since this file was already there I just missed that part, will add.

LaakkonenJussi replied:
Done, so resolving.

Tomin1 commented on permissions/DataStorages.permission:
```diff
@@ -5,3 +5,9 @@
 # x-sailjail-long-description = Use memory cards and USB sticks
+
+### Allow to listen/talk to udisks2
+# Used for mounting, formatting
```
I wonder if we should allow this at all. Then again, we need to check that UDisks.permission still anyway, so it's not too critical to do it now.

LaakkonenJussi replied:
That was what I was thinking most of the day that should some app have the permission to add/format sd/usb or is it completely internal functionality.

spiiroin replied:
Getting it off "Base" would already be a massive improvement.

But there was some reason why it got included in there... I do not remember exactly what it might have been, but typical reason for illogically lavish Base permission would be: without it some mw component crashes on unchecked access failure -> changes like this need testing / enough time in devel.

LaakkonenJussi replied:
But could we limit the scope of `UDisks2` here as well? There are some a bit dangerous D-Bus methods that can be used with this lavish rule, Delete(), Resize() do not sound like our mw apps do really need. Settings is the only one that can manage removable media and it does not yet use sailjail/firejail?

spiiroin replied:
Well, making it stricter would of course be better. With assumption like "only settings modifies" it could even be only  enumeration/getter functionality allowed or stte.

LaakkonenJussi replied:
I went through https://git.sailfishos.org/mer-core/nemo-qml-plugin-systemsettings/blob/master/src/udisks2defines.h and added the more specific interface/method rules with 3d6d780 + the `org.freedesktop.UDisks2.Manager.GetBlockDevices` which is in the code only and what has been complained in the journal if there are no UDisks2 permissions.

LaakkonenJussi replied:
I did split the permissions to `UDisks` and to `UDisksListen` - the latter of which is included by `RemovableMedia` and by `UDisks`. This way the apps can simply get info on the removable media, as well as other storage devices on the system but not to tamper with them at any level. All destructive, mounting, etc. are left in `UDisks` permission.

Tomin1 reviewed (changes requested):
Also please fill in description of DataStorages to README. I think this could be also moved from internal list to "public" list.

BTW, if you can think of a better name than DataStorages, that could be also changed now.

LaakkonenJussi commented:
> Also please fill in description of DataStorages to README. I think this could be also moved from internal list to "public" list.
> 
> BTW, if you can think of a better name than DataStorages, that could be also changed now.

Yeah well functionality first, documentation second. Early review gets the bugs etc. On it!

rainemak commented on permissions/DataStorages.permission:
```diff
@@ -5,3 +5,9 @@
+# Used for mounting, formatting
+include /etc/sailjail/permissions/UDisks.permission
+
+ignore disable-mnt
```
Wondering what else gets exposed with this other than external storages?

LaakkonenJussi replied:
No need to wonder, [the manual pages ](https://firejail.wordpress.com/features-3/man-firejail-profile/) just state this clearly:
`
       disable-mnt
              Disable /mnt, /media, /run/mount and /run/media access.
`

Unless the curtains do not match the drapes?

Tomin1 commented on permissions/UDisks.permission:
```diff
@@ -1,16 +1,18 @@
-# x-sailjail-translation-key-long-description =
-# x-sailjail-long-description =
+# x-sailjail-translation-key-long-description = permission-la-udisks2_description
+# x-sailjail-long-description = Access signals and properties on UDisks2, allows to get the devices, lock, mount and format to be called.
```
I would not mention UDisks2 here. This is user visible text (when included directly by an application) and they don't have any idea what that means.

LaakkonenJussi replied:
Yeah, that might be better now with 94588c1489e297c32c6da6ee193a4119ed181cc6 ?

Tomin1 commented on permissions/UDisksListen.permission:
```diff
@@ -0,0 +1,18 @@
+
+# x-sailjail-translation-catalog = sailfish-catalog
+# x-sailjail-translation-key-description = permission-la-udisks2_listen
+# x-sailjail-description = Use UDisks2 for listening
```
Similarly here: there shouldn't be UDisks2 here. Perhaps this shouldn't have any translations and instead it should be an internal permission that's just included from RemovableMedia.

LaakkonenJussi replied:
Better with 94588c1489e297c32c6da6ee193a4119ed181cc6 ?

Tomin1 reviewed (commented):
I left some comments to think about.

LaakkonenJussi commented on README.md:
```diff
@@ -115,6 +115,7 @@ Permissions that applications may use (names are subject to change):
 | Music | Access to Music directory, playlists and coverart cache. |
 | Phone | Make Phone calls, either directly or through system voice call UI. |
 | Pictures | Access to Pictures directory and thumbnails. |
+| RemovableMedia | Use memory cards and USB sticks |
```
Added `.` with 94588c1489e297c32c6da6ee193a4119ed181cc6

LaakkonenJussi commented:
> The format that needs to be used is:
> 
> ```
> dbus.system.[call|broadcast] <NAME>[.*]=<INTERFACE>[.*][@<OBJECT>[/*]]
> ```
> 
> There is no '=' in any of these call rules -> invalid rules for never-going-to-exist names / stte.
> 
> The documentation / even code is not always very clear, but you can do at least format checking with e.g.:
> 
> ```
> $ xdg-dbus-proxy 3>/dev/null --fd=3 fooname barsocket --call=org.freedesktop.UDisks2.Block.Format
> 'org.freedesktop.UDisks2.Block.Format' is not a valid name + rule
> ```
> 
> = not ok
> 
> ```
> $ xdg-dbus-proxy 3>/dev/null --fd=3 fooname barsocket --call=org.freedesktop.UDisks2.Block=org.freedesktop.UDisks2.Block.Format
> ^C
> ```
> 
> = ok

Thanks for clarifying my head with this as the documentation did not. Fixed these issues in the files.

Tomin1 commented on permissions/UDisks.permission:
```diff
@@ -1,16 +1,18 @@
+# x-sailjail-translation-key-description = permission-la-udisks2
+# x-sailjail-description = Manage storage devices
+# x-sailjail-translation-key-long-description = permission-la-udisks2_description
+# x-sailjail-long-description = Permission to access signals and properties on storage devices, allows to call devices get, lock, mount and format functionality.
```
I wouldn't talk about signals or properties here either because those are do not mean anything for user. These descriptions are something user sees. We can describe these separately for developer in README or here as comments.

While this is internal and will not be available in application store, this has translation catalog which means that these will generate translatable strings and translators will translate them. So we should consider that too if we add any texts here. It's also possible to leave that translation catalog empty and it will just fallback to the values here.

LaakkonenJussi replied:
Right. Lets just keep that empty and mention this only in the internals section.

LaakkonenJussi replied:
6fbe565b136b84ccbc77dc5d833ed6cdea3663fa ok?

LaakkonenJussi commented:
Updated the PR and squashed the changes to two separate commits, one for UDisks2 and one for the RemovableMedia. @Tomin1 @spiiroin @okodron I guess this is ready for final review. Checked with sailjail --trace that D-Bus log did not anymore contain any `DENIED` lines when trying to access removable media.

Tomin1 commented on permissions/UDisksListen.permission:
```diff
@@ -0,0 +1,23 @@
+# x-sailjail-translation-key-description = permission-la-udisks2_listen
+# x-sailjail-description = Access storage device information
+# x-sailjail-translation-key-long-description = permission-la-udisks2_description
+# x-sailjail-long-description = Permissions to listen signals and property changes on storage devices
```
What I said applies also here, and even more so if this is in the public list, although I would probably put this to internal list and make it available only together with RemovableMedia.

LaakkonenJussi replied:
Better to have this only with `RemovableMedia` permission, will change.

LaakkonenJussi replied:
6fbe565b136b84ccbc77dc5d833ed6cdea3663fa ok?

Tomin1 replied:
Almost, please drop those translation-catalog values still so that this won't generate translatable strings. I don't think we should have those for internal permissions unless they are intended to be used separately.

LaakkonenJussi replied:
I was about to ask some point that what is the functionality of that but it makes complete sense now. Removed.

Tomin1 replied:
Erm, sorry. That translation-catalog key doesn't affect that what I thought it was when I wrote that comment. The idea behind that key is that we can specify in permission file the translation catalog that middleware should read. However currently the only place that defines any permission files is this repository so in practice it's always the same value. Now I realized that translation generator script doesn't use that value at all (although it could check that the value is "sailjail-permissions" for correctness). So it's probably best to not to define any strings here. Alternatively we could improve the script a little.

LaakkonenJussi replied:
Dropped with bec9f7dd73014d68e81c6aa6c7f95119fe8fac45

LaakkonenJussi replied:
After a tonload of confusion and adding of 95a38b66ea3843e83dc300b4331b16a4b0ae5668 to the commits I guess it is now as it should or was intended to be.

Tomin1 replied:
I think it's great. :+1:

Tomin1 commented on permissions/RemovableMedia.permission:
```diff
@@ -0,0 +1,13 @@
+# x-sailjail-long-description = Use memory cards and USB sticks
+
+### Allow to listen/talk to udisks2
+# Used for mounting, formatting, locking only
```
Perhaps this comment should be removed still.

LaakkonenJussi replied:
Oh yeah, that is a remnant that should be removed.

LaakkonenJussi replied:
Removed by editing the relevant commit and force pushing. Resolving.

Tomin1 reviewed (approved):
I think this is fine.

LaakkonenJussi commented:
@okodron waiting for review on your side as well.

spiiroin commented on permissions/UDisks.permission:
```diff
@@ -1,16 +1,19 @@
+# BEG systembus-org.freedesktop.UDisks2 method call access
+dbus-system.call org.freedesktop.UDisks2=org.freedesktop.DBus.Properties.Set@/*
+dbus-system.call org.freedesktop.UDisks2=org.freedesktop.UDisks2.Block.Format@/org/freedesktop/UDisks2/Block
+dbus-system.call org.freedesktop.UDisks2=org.freedesktop.UDisks2.Block.Rescan@/org/freedesktop/UDisks2/Block
```
Are these object paths right?

... I have no idea, but with "qdbus --system org.freedesktop.UDisks2" I get objects like "/org/freedesktop/UDisks2/block_devices/mmcXYZ" that then do have Block interface available.

LaakkonenJussi replied:
Ah yes, maybe the interfaces could be dropped here as well for the method calls.

LaakkonenJussi replied:
Because, there are many different paths that make these same calls available.

LaakkonenJussi replied:
Like e751dcf3128d721bc18314458dcd64e7059e6c4c
